### PR TITLE
[IMP] web, *: Open link in new tab everywhere

### PR DESCRIPTION
**Pull request by jpp-odoo**
Opening link in new tab is a very cool feature for productivity purpose.

task-id: 4031939

### CODE REVIEW

**Comment by jpp-odoo** [comment]
@robodoo rebase-ff

**Comment by jpp-odoo** [comment]
@odoo/rd-security Could you override the security CI? We markup the help attribute of an action previously stored in the session storage, so it persists when open in a new tab. (commit: https://github.com/odoo/odoo/pull/185674/commits/f33a31316ec155ea76bff94075464d8b000a2477)

**Comment by alexkuhn** [comment]
Note that opening in a new tab is nice in desktop and inside browser, but when in PWA and/or in mobile, opening a new browser tab usually sucks.

I get that most users are on desktop without PWA so this is democratically a better decision. But I also want to point out a few issues with always opening links in a new tab.

**Comment by Iucapad** [comment]
> Note that opening in a new tab is nice in desktop and inside browser, but when in PWA and/or in mobile, opening a new browser tab usually sucks.
> 
> I get that most users are on desktop without PWA so this is democratically a better decision. But I also want to point out a few issues with always opening links in a new tab.

With this PR, we don't always open links in new tab. We only do, when the ctrl key/middle clicked occurs.
We just enable the feature for "any element that do a navigation", instead of just having it on <a> tags natively.

Also, PWA technology do support tabs, on some browsers (not yet standard, but some day it might make sense)

File: addons/web/static/tests/t_custom_click.test.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    await mountComponent(Root, getFixture(), { props: { text: "text from props" } });
+    after(() => {
+        delete odoo.__WOWL_DEBUG__;
+    });
+    expect(getFixture()).toHaveText("text from props");
+    await contains(".clickMe").click();
+    expect.verifySteps(["clickend on plop"]);
+});
+
+test(`detect if middle Click`, async () => {
```
Comment (jpp-odoo) [inline addons/web/static/tests/t_custom_click.test.js:31]: test real middle click (here is ctrl click)

File: addons/web/static/src/webclient/actions/action_service.js [hunk 2]
```diff
@@ -805,6 +805,21 @@ export function makeActionManager(env, router = _router) {
...
+
+            // Store current action of the current window
+            const currentAction = browser.sessionStorage.getItem("current_action");
+            // Store on the session the action for the new window
+            browser.sessionStorage.setItem("current_action", action._originalAction || "{}");
+            const href = stateToUrl(makeState(nextStack));
+            _openURL(href);
+            // restore the current action from the current window
+            browser.sessionStorage.setItem("current_action", currentAction);
+            return;
```
Comment (jpp-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:821]: refactor in function : _openActionInNewWindow(action, href);

File: addons/web/static/src/webclient/actions/action_service.js [hunk 3]
```diff
@@ -780,6 +780,27 @@ export function makeActionManager(env, router = _router) {
         return controllerStack.length;
     }
 
+    /**
+     * Open the action in a new window
+     *
+     * @param {Object} action
+     * @param {String} href
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:787]: adapt

File: addons/web/static/src/webclient/actions/action_service.js [hunk 4]
```diff
@@ -780,6 +780,27 @@ export function makeActionManager(env, router = _router) {
         return controllerStack.length;
     }
 
+    /**
+     * Open the action in a new window
+     *
+     * @param {Object} action
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:786]: ActionDescription

File: addons/web/static/tests/webclient/actions/new_window_action.test.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            type: "ir.actions.client",
+        },
+        {
+            newWindow: true,
+        }
+    );
+    expect.verifySteps(["open: /odoo/__test__client__action__"]);
+});
+
+test("retrieving the stored action when open the new window", async () => {
```
Comment (aab-odoo) [inline addons/web/static/tests/webclient/actions/new_window_action.test.js:132]: remove

File: addons/web/static/src/env.js
Comment (aab-odoo) [inline addons/web/static/src/env.js:1]: commit [IMP] web: add custom directive click

motivate

Comment (aab-odoo) [reply addons/web/static/src/env.js:1]: add link to the owl commit

File: addons/web/static/src/env.js [hunk 6]
```diff
@@ -169,6 +170,48 @@ async function _startServices(env, toStart) {
     }
 }
 
+export const customDirectives = {
+    // t-custom-click="handler"
+    // This custom directive will add two even listener ("click"; "auxclick") and call the global value "click".
```
Comment (aab-odoo) [inline addons/web/static/src/env.js:175]: listeners

File: addons/web/static/src/env.js [hunk 7]
```diff
@@ -169,6 +170,48 @@ async function _startServices(env, toStart) {
...
+    // The global value "click", will call the handler with two parameters :
```
Comment (aab-odoo) [inline addons/web/static/src/env.js:176]: "click", remove ","

File: addons/web/static/src/env.js [hunk 8]
```diff
@@ -169,6 +170,48 @@ async function _startServices(env, toStart) {
...
+    //      - ev (the original event)
+    //      - isMiddleClick (a boolean that say if the user clicked in the middle click, or if he did a ctrl+click)
```
Comment (aab-odoo) [inline addons/web/static/src/env.js:178]: says

Comment (aab-odoo) [reply addons/web/static/src/env.js:178]: if the user middle clicked

File: addons/web/static/tests/t_custom_click.test.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
+import { Component, xml } from "@odoo/owl";
+import { expect, getFixture, test } from "@odoo/hoot";
+import { contains, mountWithCleanup } from "@web/../tests/web_test_helpers";
+
+test(`simple render`, async () => {
+    class MyComponent extends Component {
+        static template = xml`<div t-custom-click="plop" class="clickMe"><t t-esc="props.text"/></div>`;
+        static props = ["*"];
+        plop() {
+            expect.step("clickend on plop");
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:10]: clicked

File: addons/web/static/tests/t_custom_click.test.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    await mountWithCleanup(MyComponent, { props: { text: "text from props" } });
+
+    expect(".clickMe").toHaveText("text from props");
+    await contains(".clickMe").click();
+    expect.verifySteps(["clickend on plop"]);
+});
+
+test(`simple render binded`, async () => {
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:21]: bound

File: addons/web/static/tests/t_custom_click.test.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
+import { Component, xml } from "@odoo/owl";
+import { expect, getFixture, test } from "@odoo/hoot";
+import { contains, mountWithCleanup } from "@web/../tests/web_test_helpers";
+
+test(`simple render`, async () => {
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:5]: standard click

File: addons/web/static/tests/t_custom_click.test.js (hunk 10 above)
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:21]: handler is bound

File: addons/web/static/tests/t_custom_click.test.js (hunk 9 above)
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:10]: check middleclick false

File: addons/web/static/tests/t_custom_click.test.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    await mountWithCleanup(MyComponent, { props: { text: "text from props" } });
+
+    expect(".clickMe").toHaveText("text from props");
+    await contains(".clickMe").click({ ctrlKey: true });
+    expect.verifySteps(["isMiddleClick: true"]);
+});
+
+test(`simple render (arrow function)`, async () => {
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:73]: rename all "simple render"

File: addons/web/static/tests/t_custom_click.test.js [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    await mountWithCleanup(MyComponent, { props: { text: "text from props" } });
+
+    expect(".clickMe").toHaveText("text from props");
+    await contains(".clickMe").click({ ctrlKey: true });
+    expect.verifySteps(["isMiddleClick: true", "text: test"]);
+});
+
+test(`"stop" modifier`, async () => {
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:127]: and prevent modifiers

File: addons/web/static/tests/t_custom_click.test.js [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    class MyComponent extends Component {
+        static template = xml`<div t-on-click="noClick"><div t-custom-click.stop="plop" class="clickMe"><t t-esc="props.text"/></div></div>`;
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:129]: .stop.prevent

File: addons/web/static/tests/t_custom_click.test.js [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        static props = ["*"];
+        plop(ev, isMiddleClick) {
+            expect.step(`isMiddleClick: ${isMiddleClick}`);
```
Comment (aab-odoo) [inline addons/web/static/tests/t_custom_click.test.js:132]: expect ev.isPreventDefaulted

File: addons/web/static/src/views/list/list_controller.js [hunk 16]
```diff
@@ -274,23 +274,28 @@ export class ListController extends Component {
         }
     }
 
-    async openRecord(record, force = false) {
+    async openRecord(record, force, newWindow) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/list_controller.js:277]: { force, newWindow }

File: addons/mail/static/src/views/web/activity/activity_controller.js [hunk 17]
```diff
@@ -117,9 +124,9 @@ export class ActivityController extends Component {
         this.model.orm.call(this.props.resModel, "activity_send_mail", [resIds, templateID], {});
     }
 
-    async openRecord(record, mode) {
+    async openRecord(record, mode, newWindow) {
```
Comment (aab-odoo) [inline addons/mail/static/src/views/web/activity/activity_controller.js:127]: { mode, newWindow }

File: addons/web/static/src/search/control_panel/control_panel.xml [hunk 18]
```diff
@@ -113,7 +113,7 @@
                                 <button class="btn btn-secondary o_switch_view "
                                     t-attf-class="o_{{view.type}} {{view.active ? 'active' : ''}}"
                                     t-att-data-tooltip="view.name"
-                                    t-on-click="() => this.switchView(view.type)"
+                                    t-custom-click="(ev, isMiddleClick) => this.switchView(view.type, isMiddleClick)"
```
Comment (aab-odoo) [inline addons/web/static/src/search/control_panel/control_panel.xml:116]: add a test

File: addons/web/static/src/env.js [hunk 19]
```diff
@@ -169,6 +170,48 @@ async function _startServices(env, toStart) {
...
+        node.setAttribute(`t-on-click${mods}`, handlerFunction);
+        node.setAttribute(`t-on-auxclick${mods}`, handlerFunction);
+    },
+};
+
+export const globalValues = {
+    click: (ev, value, modifiers) => {
+        if (ev.button === 0 || ev.button === 1) {
+            const ctrlKey = isMacOS() ? ev.metaKey : ev.ctrlKey;
+            const isMiddleClick = (ctrlKey && ev.button === 0) || ev.button === 1;
```
Comment (aab-odoo) [inline addons/web/static/src/env.js:200]: add a test for button 2

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Pull request closed without merging**